### PR TITLE
feat: support external spark

### PR DESCRIPTION
**Pull request by owlet42**
Signed-off-by: ChenLong Ma <owlet42@126.com>

### CODE REVIEW

File: docs/configurations/FATE_cluster_configuration.md [hunk 1]
```diff
@@ -152,9 +157,8 @@ Configuration of kubernetes deployment spark.
 | ----------------- | ------------ | -------- | ---------------------------- |
 | master/<br>worker | Image        | scalars  | Image of spark components    |
 |                   | ImageTag     | scalars  | ImageTag of spark components |
-|                   | replicas     |          | Number of copies of pod      |
-|                   | cpu          | scalars  | Requested CPU resources      |
-|                   | memory       | scalars  | Requested memory resources   |
+|                   | replicas     | scalars  | Number of copies of pod      |
+|                   | resources    | mappings | resources of Kubernetes      |
```
Comment (jiahaoc1993) [inline docs/configurations/FATE_cluster_configuration.md:161]: resources of Kubernetes => resource of pod ?

File: helm-charts/FATE/templates/hdfs-nn.yaml [hunk 2]
```diff
@@ -168,6 +168,9 @@ spec:
       port: 9000
       targetPort: 9000
       protocol: TCP
+      {{- if eq .Values.modules.hdfs.namenode.type "NodePort" "LoadBalancer" }}
```
Comment (jiahaoc1993) [inline helm-charts/FATE/templates/hdfs-nn.yaml:171]: why mixing `nodeport` and `loadbalancer` here?

File: helm-charts/FATE/templates/python-spark.yaml [hunk 3]
```diff
@@ -426,6 +434,40 @@ spec:
...
+{{ include "fate.labels" . | indent 4 }}
+spec:
+  ports:
+    {{ $spark := .Values.modules.python.spark }}
+    {{- range .Values.modules.python.spark.portMaxRetries | int | until }}
+    - name: "spark-driver-port-{{ . }}"
+      port: {{ $spark.driverStartPort | int | add . }}
+      targetPort: {{ $spark.driverStartPort | int | add .  }}
+      {{- if eq $spark.driverHostType "NodePort" "LoadBalancer" }}
+      nodePort: {{ $spark.driverStartPort | int | add .  }}
```
Comment (jiahaoc1993) [inline helm-charts/FATE/templates/python-spark.yaml:453]: I don't think `LoadBalancer` type of service requires this `nodeport` field

File: docs/configurations/FATE_cluster_configuration.md [hunk 4]
```diff
@@ -107,20 +107,25 @@ The parties are directly connected.
...
+| spark                 | mappings | If you use your own spark, modify the configuration          |
+| spark.cores_per_node  | scalars  | configuration of FATE fateflow module                        |
+| spark.nodes           | scalars  | configuration of FATE fateflow module                        |
+| spark.existingSpark   | scalars  | If you need to use the existing spark , you can set this configuration |
+| spark.driverHost      | scalars  | call back IP of spark executor                               |
+| spark.driverHostType  | scalars  | service type of spark driver                                 |
+| spark.portMaxRetries  | scalars  | spark driver's configuration                                 |
+| driverStartPort       | scalars  | spark driver start port                                      |
+| blockManagerStartPort | scalars  | spark driver blockManager start port                         |
+| pysparkPython         | scalars  | spark worker node python PATH                                |
```
[the remark this answers is not in the document]
Comment (owlet42) [reply docs/configurations/FATE_cluster_configuration.md]: yes, it was changed

File: helm-charts/FATE/values-template-example.yaml [hunk 5]
```diff
@@ -120,26 +120,28 @@ backend: eggroll
   # accessMode: ReadWriteMany
   # size: 1Gi
   # spark: 
-    # master: spark://spark-master:7077
-    # home: 
     # cores_per_node: 20
     # nodes: 2
+    # existingSpark: spark://192.168.99.1:7077
+    # driverHost: fateflow
```
Comment (jiahaoc1993) [inline helm-charts/FATE/values-template-example.yaml:126]: rename "fateflow" to "<<LB Address>> or <<WorkerIP>>"

File: helm-charts/FATE/templates/hdfs-nn.yaml (hunk 2 above)
Comment (owlet42) [reply helm-charts/FATE/templates/hdfs-nn.yaml:171]: We can deploy a hdfs cluster, the entrance use 'NodePort' or 'LoadBalancer'.

File: helm-charts/FATE/values-template-example.yaml [hunk 6]
```diff
@@ -120,26 +120,28 @@ backend: eggroll
...
+    # driverHostType: NodePort
+    # portMaxRetries: 16
+    # driverStartPort: 30997
+    # blockManagerStartPort: 31107
+    # pysparkPython: 
   # hdfs:
-    # name_node: hdfs://namenode:9000
+    # existingHdfs: hdfs://namenode:9000
```
Comment (jiahaoc1993) [inline helm-charts/FATE/values-template-example.yaml:133]: why not use `name_node`?

File: helm-charts/FATE/templates/python-spark.yaml (hunk 3 above)
Comment (owlet42) [reply helm-charts/FATE/templates/python-spark.yaml:453]: If LoadBalancer does not have a definite IP, sometimes nodeport is a better way.

File: helm-charts/FATE/values-template-example.yaml [hunk 7]
```diff
@@ -120,26 +120,28 @@ backend: eggroll
   # accessMode: ReadWriteMany
   # size: 1Gi
   # spark: 
-    # master: spark://spark-master:7077
-    # home: 
     # cores_per_node: 20
     # nodes: 2
+    # existingSpark: spark://192.168.99.1:7077
```
Comment (jiahaoc1993) [inline helm-charts/FATE/values-template-example.yaml:125]: I think it should change back to "master"

File: helm-charts/FATE/templates/python-spark.yaml [hunk 8]
```diff
@@ -19,10 +19,18 @@ metadata:
 {{ include "fate.labels" . | indent 4 }}
 data:
   spark-defaults.conf: |
-    spark.master                      {{  .Values.modules.python.spark.master | default "local[4]"}}
-    #spark.eventLog.enabled           true
-    #spark.eventLog.dir               hdfs://namenode:9000/spark/logs
-    spark.driver.host                 fateflow
+    spark.master                      {{  .Values.modules.python.spark.existingSpark | default "spark://spark-master:7077"}}
```
Comment (jiahaoc1993) [inline helm-charts/FATE/templates/python-spark.yaml:22]: default value is splitting everywhere, could you please unify them in a single place?

File: helm-charts/FATE/values-template-example.yaml (hunk 6 above)
Comment (owlet42) [reply helm-charts/FATE/values-template-example.yaml:133]: You only need to set it when you use an existing HDFS.

**Pull request closed without merging**